pkg/solana/chainwriter/lookups.go

```diff
@@ -0,0 +1,356 @@
+		}
+
+		result = append(result, &solana.AccountMeta{
+			PublicKey:  solana.PublicKeyFromBytes(value[0]),
+			IsSigner:   accountMeta.IsSigner,
+			IsWritable: accountMeta.IsWritable,
+		})
+	}
+	return result, nil
+}
+
+func decodeBorshIntoType(data []byte, typ reflect.Type) (interface{}, error) {
+	// Ensure the type is a struct
+	if typ.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("provided type is not a struct: %s", typ.Kind())
+	}
+
+	// Create a new instance of the type
+	instance := reflect.New(typ).Interface()
+
+	// Decode using Borsh
+	err := ag_binary.NewBorshDecoder(data).Decode(instance)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding Borsh data: %w", err)
+	}
+
+	// Return the underlying value (not a pointer)
+	return reflect.ValueOf(instance).Elem().Interface(), nil
+}
+
+// getSeedBytes extracts the seeds for the PDALookups.
+// It handles both AddressSeeds (which are public keys) and ValueSeeds (which are byte arrays from input args).
+func getSeedBytes(ctx context.Context, lookup PDALookups, args any, derivedTableMap map[string]map[string][]*solana.AccountMeta, reader client.Reader) ([][]byte, error) {
+	var seedBytes [][]byte
+
+	for _, seed := range lookup.Seeds {
+		if lookupSeed, ok := seed.(AccountLookup); ok {
+			// Get value from a location (This doens't have to be an address, it can be any value)
+			bytes, err := GetValuesAtLocation(args, lookupSeed.Location)
+			if err != nil {
+				return nil, fmt.Errorf("error getting address seed: %w", err)
+			}
+			seedBytes = append(seedBytes, bytes...)
+		} else {
+			// Get address seeds from the lookup
+			seedAddresses, err := GetAddresses(ctx, args, []Lookup{seed}, derivedTableMap, reader)
+			if err != nil {
+				return nil, fmt.Errorf("error getting address seed: %w", err)
+			}
+
+			// Add each address seed as bytes
+			for _, address := range seedAddresses {
+				seedBytes = append(seedBytes, address.PublicKey.Bytes())
+			}
+		}
+	}
+
+	return seedBytes, nil
+}
+
+// generatePDAs generates program-derived addresses (PDAs) from public keys and seeds.
+func generatePDAs(publicKeys []*solana.AccountMeta, seeds [][]byte, lookup PDALookups) ([]*solana.AccountMeta, error) {
+	if len(seeds) > 1 && len(publicKeys) > 1 {
+		return nil, fmt.Errorf("multiple public keys and multiple seeds are not allowed")
+	}
+
+	var addresses []*solana.AccountMeta
+	for _, publicKeyMeta := range publicKeys {
+		address, _, err := solana.FindProgramAddress(seeds, publicKeyMeta.PublicKey)
+		if err != nil {
+			return nil, fmt.Errorf("error finding program address: %w", err)
+		}
+		addresses = append(addresses, &solana.AccountMeta{
+			PublicKey:  address,
+			IsSigner:   lookup.IsSigner,
+			IsWritable: lookup.IsWritable,
+		})
+	}
+	return addresses, nil
+}
+
+func (s *SolanaChainWriterService) ResolveLookupTables(ctx context.Context, args any, lookupTables LookupTables) (map[string]map[string][]*solana.AccountMeta, map[solana.PublicKey]solana.PublicKeySlice, error) {
```

Comment:
Yeah, I think it makes sense to include these in `chain_writer.go` as opposed to `lookups.go`. I moved them over.

Reply:
> Also, are both of these functions internal to this struct? If yes, can we not export them? Otherwise it appears to the reader that they are used from outside too.
I made `loadTable` unexported, however, ResolveLookupTables is used outside of the package in [lookups_test.go](https://github.com/smartcontractkit/chainlink-solana/blob/54e3ae30a247dfc24d3413d1c6ce8a3265558109/integration-tests/relayinterface/lookups_test.go#L307). I think it makes sense to have that test be external since it is an integration test which uses the Solana test validator